Use of rsync for local repository

I have a local repository (a local checkout of the coq/repo-stable repository):

    $ opam repo
      10 [local]   coq-repo     /home/guillaume/src/opam-coq-repo
       0 [http]     default     https://opam.ocaml.org

I already have used this setup successfully some times ago (but perhaps not with opam 1.2), but now when I try to install a package described in this repository, I get the following behavior:

    $ opam install coq:interval
    The following actions will be performed:
     - install   coq:flocq.2.4.0                       [required by coq:interval]
     - install   coq:ssreflect.1.5.0                   [required by coq:interval]
     - install   coq:interval.2.0.0
    === 3 to install ===
    Do you want to continue ? [Y/n] 
    
    =-=- Synchronizing package archives -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
    [WARNING] Rsync partially failed:
      rsync: link_stat "/home/guillaume/src/opam-coq-repo/archives/coq:interval.2.0.0+opam.tar.gz" failed: No such file or directory (2)
      rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1183) [sender=3.1.1]
    [WARNING] Rsync partially failed:
      rsync: link_stat "/home/guillaume/src/opam-coq-repo/archives/coq:ssreflect.1.5.0+opam.tar.gz" failed: No such file or directory (2)
      rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1183) [sender=3.1.1]
    [WARNING] Rsync partially failed:
      rsync: link_stat "/home/guillaume/src/opam-coq-repo/archives/coq:flocq.2.4.0+opam.tar.gz" failed: No such file or directory (2)
      rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1183) [sender=3.1.1]
    [ERROR] Could not download archives of coq:flocq.2.4.0, coq:interval.2.0.0 and coq:ssreflect.1.5.0

When using the --debug flag, I see that opam tried to use rsync:

    rsync_file src=/home/guillaume/src/opam-coq-repo/archives/coq:flocq.2.4.0+opam.tar.gz dst=/home/guillaume/.opam/repo/coq-repo/archives/coq:flocq.2.4.0+opam.tar.gz

instead of downloading the tarballs described in the repository.

This is probably not a bug, probably just a misconfiguration, but the documentation was not helpful in understanding why rsync is used rather than curl for this local repository.

--- a/src/client/opamState.ml
+++ b/src/client/opamState.ml
@@ -2744,7 +2744,7 @@ let download_archive t nv =
         OpamGlobals.msg "%s Repo archive not found" text;
       Done None
     | Up_to_date f ->
-      OpamGlobals.msg "%s Archive in cache" text;
+      OpamGlobals.msg "%s Archive in cache\n" text;
       OpamFilename.copy ~src:f ~dst; Done (Some dst)
     | Result f ->
       OpamFilename.copy ~src:f ~dst; Done (Some dst)

--- a/src/repositories/opamLocal.ml
+++ b/src/repositories/opamLocal.ml
@@ -23,7 +23,7 @@ let slog = OpamGlobals.slog
 
 let rsync_arg = "-rLptgoDrvc"
 
-let call_rsync args =
+let call_rsync check args =
   OpamSystem.make_command "rsync" args
   @@> fun r ->
   match r.OpamProcess.r_code with
@@ -32,10 +32,14 @@ let call_rsync args =
     Done None
   | 20 -> (* signal *)
     raise Sys.Break
-  | 23 | 24 -> (* partial, mostly mode, link or perm errors *)
-    OpamGlobals.warning "Rsync partially failed:\n  %s"
-      (String.concat "\n  " r.OpamProcess.r_stderr);
-    Done (Some (OpamMisc.rsync_trim r.OpamProcess.r_stdout))
+  | 23 | 24 ->
+    (* partial, mostly mode, link or perm errors. But may also be a
+       complete error so we do an additional check *)
+    if check () then
+      (OpamGlobals.warning "Rsync partially failed:\n  %s"
+         (String.concat "\n  " r.OpamProcess.r_stderr);
+       Done (Some (OpamMisc.rsync_trim r.OpamProcess.r_stdout)))
+    else Done None
   | 30 | 35 -> (* timeouts *)
     Done None
   | _ -> OpamSystem.process_error r
@@ -49,13 +53,14 @@ let rsync ?(args=[]) src dst =
     Done (Up_to_date [])
   else (
     OpamSystem.mkdir dst;
-    call_rsync ( rsync_arg :: args @ [
-                 "--exclude"; ".git";
-                 "--exclude"; "_darcs";
-                 "--exclude"; ".hg";
-                 "--exclude"; ".#*";
-                 "--delete";
-                 src; dst; ])
+    call_rsync (fun () -> not (OpamSystem.dir_is_empty dst))
+      ( rsync_arg :: args @ [
+            "--exclude"; ".git";
+            "--exclude"; "_darcs";
+            "--exclude"; ".hg";
+            "--exclude"; ".#*";
+            "--delete";
+            src; dst; ])
     @@| function
     | None -> Not_available src
     | Some [] -> Up_to_date []
@@ -87,7 +92,8 @@ let rsync_file ?(args=[]) src dst =
   else if src_s = dst_s then
     Done (Up_to_date src)
   else
-  call_rsync ( rsync_arg :: args @ [ src_s; dst_s ])
+  call_rsync (fun () -> Sys.file_exists dst_s)
+    ( rsync_arg :: args @ [ src_s; dst_s ])
   @@| function
   | None -> Not_available src_s
   | Some [] -> Up_to_date dst

